source/agora/common/Config.d

```diff
@@ -81,1059 +136,693 @@ public struct CommandLine
-    Duration timeout = get!(Duration, "node", "timeout", str => str.to!ulong.msecs)
-        (cmdln, node);
-
-    string data_dir = get!(string, "node", "data_dir")(cmdln, node);
-    const stats_listening_port = opt!(ushort, "node", "stats_listening_port")(cmdln, node);
-    const block_time_offset_tolerance_secs = opt!(uint, "node", "block_time_offset_tolerance_secs")(cmdln, node);
-    const network_discovery_interval = opt!(uint, "node", "network_discovery_interval_secs")(cmdln, node);
-    const block_catchup_interval = opt!(uint, "node", "block_catchup_interval_secs")(cmdln, node);
-    const double_spent_threshold_pct = opt!(ubyte, "node", "double_spent_threshold_pct")(cmdln, node);
-    const relay_tx_max_num = opt!(ushort, "node", "relay_tx_max_num")(cmdln, node, 100);
-    Duration relay_tx_interval = opt!(ulong, "node", "relay_tx_interval_secs")(cmdln, node, 15).seconds;
-    const relay_tx_min_fee = Amount(opt!(ulong, "node", "relay_tx_min_fee")(cmdln, node, 0));
-    Duration relay_tx_cache_exp = opt!(ulong, "node", "relay_tx_cache_exp_secs")(cmdln, node, 1200).seconds;
-
-    NodeConfig result = {
-            min_listeners : min_listeners,
-            max_listeners : max_listeners,
-            commons_budget_address : commons_budget_address,
-            testing : testing,
-            limit_test_validators: limit_test_validators,
-            retry_delay : retry_delay,
-            max_retries : max_retries,
-            timeout : timeout,
-            data_dir : data_dir,
-            stats_listening_port : stats_listening_port,
-            block_time_offset_tolerance : block_time_offset_tolerance_secs.seconds,
-            network_discovery_interval : network_discovery_interval.seconds,
-            block_catchup_interval : block_catchup_interval.seconds,
-            double_spent_threshold_pct : double_spent_threshold_pct,
-            relay_tx_max_num : relay_tx_max_num,
-            relay_tx_interval : relay_tx_interval,
-            relay_tx_min_fee : relay_tx_min_fee,
-            relay_tx_cache_exp : relay_tx_cache_exp,
-    };
-    return result;
-}
+/// Convenience short-hand template to get a field identifier
+private enum FId (alias Field) = __traits(identifier, Field);
 
-///
-unittest
-{
-    import dyaml.loader;
+/// Evaluates to `true` if this field is to be considered optional
+/// (does not need to be present in the YAML document)
+private enum isOptional (alias FR) = hasUDA!(FR.Ref, Optional) ||
+    is(immutable(FR.Type) == immutable(bool)) ||
+    (FR.Default != FR.Type.init);
 
-    CommandLine cmdln;
+/// Evaluates to `true` if we should recurse into the struct via `parseDefaultMapping`
+private enum mightBeOptional (alias FR) = is(FR.Type == struct) &&
+    !hasConverter!(FR.Ref) && !hasFromString!(FR.Type) && !hasStringCtor!(FR.Type);
 
-    {
-        immutable conf_example = `
-node:
-  interfaces:
-    - type: http
-      address: 0.0.0.0
-      port: 2926
-  data_dir: .cache
```

Comment:
Strict mode blew a fuse on this, so I have removed the nested `interfaces` as it is incorrect.